Feature/job groups

dylandreimerink opened the pull request:
This PR introduces a new job group system. Up to this point we did not have a "good" way to manage jobs/work/routines in hive cells. While the lifecycle Start and Stop hook methods are really great for guaranteeing ordered startup and shutdown, they are not great on the usage side. Most code expect to have a context and just stop when it cancels, which is reasonable. However converting between the two paradigms takes a lot of boilerplate code: a waitgroup, context and cancel function. 

Until now we have been using `cilium/workerpool` to reduce the manual work, however, workerpool requires you to specify the amount of goroutines upfront, and submitting more work than workers will queue work. It is simply not designed for what we are doing with it. This job system is designed to replace both of these older methods.

In the pre-hive code it is common to see the pkg/controller used. It did what lifecycles do for us now but with some extra features such as retries, periodic invocation and triggers. It also monitors the controllers and makes that info available as metrics. In anticipation of replacing the current controller system with this new job group system, I have added a few features which are present in pkg/controllers so future migration of code that uses controllers is easier.

```release-note
Added a new job group system to manage the lifecycle of jobs within cells
```

dylandreimerink committed:
hive: Add job groups

This commit introduces a new package call `job`. Which provides a
`job.Collection` and `job.Group`. The collection is created once as a
cell in the hive. Other cells will then take the collection as a input
and make a new Group from the collection and assign it to their own
cell structure.

The group has a number of function to add different job types:

* AddOneShot - Adds a 'one shot' job which runs once, can exit early or
  remain running for the entire life of the cell. On error, the job can
  be retried.
* AddTimer - Adds a timer job which runs at an given interval or
  optionally when triggered.
* AddObserver - Adds a observer job which triggers on every item sent
  over the given stream.Observable.

The group implements hive.HookInterface and thus can be added directly
to a lifecycle. The jobs are invoked with contexts linked to the
cells lifecycle. The group ensured proper shutdown behavior and job
scheduling with minimal boilerplate for users.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

tklauser reviewed (commented):
Haven't looked at the implementation yet. I noticed that the job groups API isn't documented as part of the hive API in [`Documentation/contributing/development/hive.rst`](https://github.com/cilium/cilium/blob/master/Documentation/contributing/development/hive.rst#the-hive-api) in this PR, but it probably should be.

joamaki commented on pkg/hive/job/group.go:
```diff
@@ -0,0 +1,473 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package job
```
Would `jobgroup` be a better package name?

dylandreimerink replied:
Yea, I didn't do that at first since `jobgroup.Group` stutters, so `job.Group` and `job.Collection` seemed more natural. But I don't mind `jobgroup` that much.

joamaki commented on pkg/hive/job/group.go:
```diff
@@ -0,0 +1,473 @@
+
+// A Collection creates Groups, it maintains references to these groups for the purposes of collecting information
+// centralized like metrics.
+type Collection struct {
```
Hm, somehow the naming doesn't quite work. `Collection` doesn't quite capture that this thing is the one creating groups.

Would any of these work better?

```
  var reg jobgroup.Registry

  jobs := reg.NewGroup(some, option)
  jobs.AddOneShot(...)  
```
```
  var provider jobgroup.Provider
  jobs := provider.NewGroup(...)
```

Or maybe making it hierarchical and inheriting options?
```
  package jobgroup

  type RootGroup *jobgroup.Group
  var Cell = cell.Provide(newRootGroup)

  package foo

  var rootGroup jobgroup.RootGroup
  jobs := rootGroup.NewGroup(...)
```
  
I would make `Group` an interface and document the different ways of adding jobs clearly in the type definition so it's easy to use this and we don't need to worry about anyone calling unwanted methods like Start and Stop.

Also all of this needs to have tests before we merge this.

dylandreimerink replied:
Yea, I like `Registry`. That is a more accurate description of its purpose anyway. Making the group an interface also seems like a good idea.

joamaki reviewed (changes requested):
Great start! I think it makes sense to not add all the bells and whistles into this in one PR, but rather try to keep it minimal like it is. What I'd like to see is good godocs (strongly consider interface types for the API to minimize the visible API surface), decent tests of all different job types, and maybe think a bit more about the naming.

joamaki commented on operator/pkg/lbipam/cell.go:
```diff
@@ -29,8 +30,9 @@ type LBIPAMParams struct {
-	Shutdowner hive.Shutdowner
+	LC            hive.Lifecycle
+	Shutdowner    hive.Shutdowner
+	JobCollection job.Registry
```
s/Collection/Registry/

joamaki commented on operator/cmd/root.go:
```diff
@@ -106,6 +107,9 @@ var (
 			}
 		}),
 
+		// Provides a global job collection which cells can use to spawn job groups.
```
s/collection/registry/. The comment feels very jargony. Wonder if it's purpose could be explained more clearly?

joamaki commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -84,9 +90,17 @@ func newLBIPAM(params LBIPAMParams) *LBIPAM {
 
-	params.LC.Append(lbIPAM)
+	jobGroup.Add(
+		job.OneShot("lb-ipam main", func(ctx context.Context) error {
```
s/lb-ipam/lbipam/ to match with the spelling above in pprof labels?

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+var Cell = cell.Provide(newRegistry)
```
Here we could add an extensive godoc comment to explain what this cell is for, what the API is and an example on how to use it. Check that it looks good with "go doc job"

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+
+	g := &group{
+		options: options,
+
```
unnecessary empty line

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	hive.HookInterface
+}
+
+type Job interface {
```
comment here on what a "Job" is

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	jg.cancel()
+
+	select {
+	case <-stopCtx.Done():
```
I'm still not quite sure how we should handle the stop context. It feels wrong to silently give up and leave the resources around. It's a bit hard to say whether failing the stop when we couldn't clean up is less safe. Alternative is to ignore the stop context (as we're not doing a long running operation that we could "safely" abort) and just wait. If we overrun the timeout hive will forcefully stop. We probably should pick what semantics to recommend, but not quite sure yet since it somewhat depends on what other stop hooks might be doing and how critical it is that they run.

For now, let's at least log that we didn't manage to wait for jobs to stop and we're leaving them running.

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	defer jg.mu.Unlock()
+
+	for _, j := range jobs {
+		if jg.ctx == nil {
```
Add a comment here that we're using the `ctx` as proxy for whether we've started yet or not.

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+		}
+	}
+
+	fields := logrus.Fields{
```
Consider:

```
l = l.WithFields(logrus.Fields{...})
```

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	for i := 0; i <= jos.retry; i++ {
+		if i != 0 {
+			timeout := jos.backoff.When(jos)
+			l.WithFields(logrus.Fields{
```
This lost the `fields`. See comment above for a solution :)

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+		}
+	}
+
+	fields := logrus.Fields{
```
ditto: `l = l.WithFields(...)`

joamaki commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+		select {
+		case <-triggerChan:
+		case <-timer.C:
+		case <-ctx.Done():
```
maybe shuffle the cases so that `<-ctx.Done()` is first. For an old C hacker like me I'm still having hard time not reading this as all falling through to `return nil`.

joamaki commented on pkg/hive/job/job_test.go:
```diff
@@ -0,0 +1,757 @@
+		l.Append(g)
+	})
+
+	if err := h.Start(context.Background()); err != nil {
```
Consider using https://pkg.go.dev/github.com/stretchr/testify/assert. 
```
if assert.NoError(t, h.Start(context.Background()) {
  <-stop
  assert.NoError(t, h.Stop(context.Background())
}
```

joamaki commented on pkg/hive/job/job_test.go:
```diff
@@ -0,0 +1,757 @@
+	for i := 1; i < len(times); i++ {
+		diff := times[i].Sub(times[i-1])
+		if i > 2 {
+			// Test that the rate of change is 2 +- 50%, the 50% to account for CI time dilation.
```
A bit torn here.I've had to remove checks like this in the past as I couldn't make them reliable enough, even with plenty of headroom. I'm pretty sure this will end up being flaky as VMs often have huge time dilation.

How about writing all the tests here that deal with time to retry multiple times on failures? This way a temporary "time dilation" won't screw us over.

tommyp1ckles commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+
+	for _, j := range jobs {
+		if jg.ctx == nil {
+			jg.queuedJobs = append(jg.queuedJobs, j)
```
Do we need these after the initial Start? Could we set the slice ref to be nil so they can be GC'd?

tommyp1ckles commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,485 @@
+	jg.wg.Add(len(jg.queuedJobs))
+	for _, job := range jg.queuedJobs {
+		pprof.Do(jg.ctx, jg.options.pprofLabels, func(ctx context.Context) {
+			go job.start(ctx, jg.wg, jg.options)
```
job.start returns an error, but we never collect it. Is that intended?

dylandreimerink replied:
No forgot about that one, thx. Good catch

tommyp1ckles commented on operator/pkg/lbipam/lbipam.go:
```diff
@@ -71,6 +72,11 @@ func newLBIPAM(params LBIPAMParams) *LBIPAM {
 
+	jobGroup := params.JobCollection.NewGroup(
+		job.WithLogger(params.Logger),
+		job.WithPprofLabels(pprof.Labels("cell", "lbipam")),
```
Just an idea, would it be possible to inject this label from the Cell automatically?

dylandreimerink replied:
I did think about this. It could be possible but only if we integrate this more tightly with the hive package. Can't do it as a separate package since that would cause an import cycle.

tommyp1ckles replied:
Makes sense,  it would just be useful as a stand base set of labels.

dylandreimerink committed:
lb-ipam: Replace worker group with job group

LB-IPAM is an easy target to implement the new job group since it is
well tested and is a drop in replacement.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

thorn3r commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,480 @@
+	defer jg.mu.Unlock()
+
+	for _, j := range jobs {
+		// The context is only set once the group has been stared. If we have not yet stared, queue the job.
```
```suggestion
		// The context is only set once the group has been started. If we have not yet started, queue the job.
```
nit

thorn3r commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,480 @@
+// once at startup. It can live for the entire lifetime of the group or exit early depending on its task.
+// If it returns an error, it can optionally be retried if the WithRetry option. If retries are not configured or
+// all retries failed as well, a shutdown of the hive can be triggered by specifying the WithShutdown option is
+// specified.
```
```suggestion
// all retries failed as well, a shutdown of the hive can be triggered by specifying the WithShutdown option.
```
nit

thorn3r commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,480 @@
+			timeout := jos.backoff.When(jos)
+			l.WithFields(logrus.Fields{
+				"backoff":     timeout,
+				"retry-count": i + 1,
```
probably being a bit pedantic here, but this 1st time this logs it will have "retry-count=2" on the 1st retry. maybe leaving this as `i` or s/retry-count/attempts/ might be a bit clearer

thorn3r commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,480 @@
+	c chan struct{}
+}
+
+func (t *trigger) Trigger() {
+	select {
+	case t.c <- struct{}{}:
+	default:
+	}
+}
```
this is slick :+1:

thorn3r commented on pkg/hive/job/job_test.go:
```diff
@@ -0,0 +1,769 @@
+	"runtime"
+	"testing"
+	"time"
+
```
linting error down here

thorn3r reviewed (approved):
this is really cool, great work. left a couple small nits

tklauser commented on pkg/hive/job/job.go:
```diff
@@ -0,0 +1,479 @@
+func (jg *group) startJob(job Job, ctx context.Context) {
+	err := job.start(ctx, jg.wg, jg.options)
+	if err != nil {
+		jg.options.logger.WithError(err).Error("Job start returned an error")
```
Do we need to log that error here. It seems it's already logged in all `.start` methods. I'd say that logging once either here on in `.start` is probably enough.

We instead also could bubble up the error to the callers, combine them there and return them to their respective callers instead of logging them in this package.

dylandreimerink replied:
We can't report the errors back to the caller since the actual job start is performed in a goroutine and might be queued. I have removed this central logging point and ensured that the `job.Start` implementations all log any errors internally.

I have also added a section to the "Guide to Hive" docs.

tklauser replied:
:+1: thanks for clarifying.

tklauser reviewed (changes requested):
Would be nice to have the _Guide to the Hive_ docs as part of this PR, ref. https://github.com/cilium/cilium/pull/24558#pullrequestreview-1360607122

dylandreimerink committed:
docs: Updated the "Guide to Hive" docs with a section about jobs

This section describes the purpose of the package and contains a
comprehensive example of possible uses of the jobs.

Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>

zacharysarah commented on Documentation/contributing/development/hive.rst:
```diff
@@ -877,4 +877,150 @@ See also the runnable example in `pkg/k8s/resource/example <https://github.com/c
+
+The `job package <https://pkg.go.dev/github.com/cilium/cilium/pkg/hive/job>`_ contains logic that 
+makes it easy to manage units of work that the package refers to as "jobs". These jobs are 
+scheduled as part of a job group. These jobs themselves come in a variety of flavors.
```
For better localization:
```suggestion
scheduled as part of a job group. These jobs themselves come in several varieties.
```

zacharysarah commented on Documentation/contributing/development/hive.rst:
```diff
@@ -877,4 +877,150 @@ See also the runnable example in `pkg/k8s/resource/example <https://github.com/c
+makes it easy to manage units of work that the package refers to as "jobs". These jobs are 
+scheduled as part of a job group. These jobs themselves come in a variety of flavors.
+
+Every job, fundamentally is a callback function provided by the user with additional logic which
+is slightly different for each job type. The jobs and groups manage a lot of the boilerplate
+surrounding lifecycle management. The callbacks are called from the job to perform the actual
+work.
+
+Take the following somewhat contrived example:
```
Grammar, clarity
```suggestion
Every job is a callback function provided by the user with additional logic which
differs slightly for each job type. The jobs and groups manage a lot of the boilerplate
surrounding lifecycle management. The callbacks are called from the job to perform the actual
work.

Consider the following example:
```

zacharysarah commented on Documentation/contributing/development/hive.rst:
```diff
@@ -877,4 +877,150 @@ See also the runnable example in `pkg/k8s/resource/example <https://github.com/c
+    }
+
 
+The above example shows a number of use-cases in one cell. We start by requesting the job.Registry
+via the constructor. We can use the registry to create job groups, in most cases one will be enough.
+To this group we can add our jobs in the constructor. Any jobs added in the constructor are queued
+until the lifecycle of our cell starts. The group is added to the lifecycle and manages this 
+internally. Jobs can also be added at runtime which can be handy for dynamic workloads while still
+guaranteeing a clean shutdown.
```
Grammar, clarity, use active voice in present tense, avoid "we" and focus instead on a reader's actions ("you" often works)
```suggestion
The preceding example shows a number of use cases in one cell. The cell starts by requesting the job.Registry
by way of the constructor. The registry can create job groups; in most cases, one is enough.
You can add jobs in the constructor to this group. Any jobs added in the constructor are queued
until the lifecycle of the cell starts. The group is added to the lifecycle and manages jobs 
internally. You can also add jobs at runtime, which can be handy for dynamic workloads while still
guaranteeing a clean shutdown.
```

zacharysarah commented on Documentation/contributing/development/hive.rst:
```diff
@@ -877,4 +877,150 @@ See also the runnable example in `pkg/k8s/resource/example <https://github.com/c
+internally. Jobs can also be added at runtime which can be handy for dynamic workloads while still
+guaranteeing a clean shutdown.
+
+A job group will cancel the context to all jobs when the lifecycle ends. Any job callbacks are 
+expected to exit as soon as possible when the ``ctx`` is "Done". The group will make sure that all
+jobs are properly shutdown before the cell stops. If callbacks that do not stop within reasonable 
+amount of time may cause hive to perform a hard shutdown.
```
Use present tense, grammar
```suggestion
A job group cancels the context to all jobs when the lifecycle ends. Any job callbacks are 
expected to exit as soon as the ``ctx`` is "Done". The group makes sure that all
jobs are properly shut down before the cell stops. Callbacks that do not stop within a reasonable 
amount of time may cause the hive to perform a hard shutdown.

zacharysarah commented on Documentation/contributing/development/hive.rst:
```diff
@@ -877,4 +877,150 @@ See also the runnable example in `pkg/k8s/resource/example <https://github.com/c
+jobs are properly shutdown before the cell stops. If callbacks that do not stop within reasonable 
+amount of time may cause hive to perform a hard shutdown.
+
+There are 3 job types: one-shot jobs, timer jobs, and observer jobs. One shot jobs run a limited 
+amount of times, they can be used for short running jobs or jobs that span the entire lifecycle.
+Once the callback exits without error, its never called again. A one-shot can optionally have retry
+logic and/or trigger hive shutdown if it fails. Timers are called on a specified interval but they
+can also be externally triggered. Lastly, we have observer jobs which are invoked for every event
+on a ``stream.Observable``.
```
Consistent hyphens, grammar
```suggestion
There are 3 job types: one-shot jobs, timer jobs, and observer jobs. One-shot jobs run a limited 
number of times: you can use them for brief jobs, or for jobs that span the entire lifecycle.
Once the callback exits without error, it is never called again. Optionally, a one-shot job can include retry
logic and/or trigger hive shutdown if it fails. Timers are called on a specified interval but they
can also be externally triggered. Lastly, observer jobs are invoked for every event
on a ``stream.Observable``.
```

joestringer replied:
@zacharysarah thoughts on "use them" vs. "you can use them"? The former seems more concise and imperative.

(Otherwise all of this feedback looks good, @dylandreimerink / @zacharysarah do you think either of you would be able to apply these diffs in a fresh PR to the tree?)

zacharysarah replied:
@joestringer Sorry, I'm just now seeing this comment. You could use either, but "use them" is totally appropriate here, and I kinda like it.

zacharysarah reviewed (commented):
@dylandreimerink 👋🏻 I see this review is post-merge, but here are my suggestions for improving content. You have a good conversational style. ✨
